unison-cli/src/Unison/Cli/Monad.hs

```diff
@@ -0,0 +1,349 @@
+    loopStateRef :: IORef LoopState,
+    -- | What to do with output for the user.
+    notify :: Output -> IO (),
+    -- | What to do with numbered output for the user.
+    notifyNumbered :: NumberedOutput -> IO NumberedArgs,
+    runtime :: Runtime Symbol,
+    sandboxedRuntime :: Runtime Symbol,
+    serverBaseUrl :: Maybe Server.BaseUrl,
+    ucmVersion :: UCMVersion
+  }
+  deriving stock (Generic)
+
+-- | The command-line app monad mutable state.
+--
+-- Get the state with 'getLoopState', put it with 'putLoopState', or modify it with 'modifyLoopState'.
+--
+-- There's an additional pseudo @"currentPath"@ field lens, for convenience.
+data LoopState = LoopState
+  { root :: Branch IO,
+    lastSavedRoot :: Branch IO,
+    -- the current position in the namespace
+    currentPathStack :: List.NonEmpty Path.Absolute,
+    -- TBD
+    -- , _activeEdits :: Set Branch.EditGuid
+
+    -- The file name last modified, and whether to skip the next file
+    -- change event for that path (we skip file changes if the file has
+    -- just been modified programmatically)
+    latestFile :: Maybe (FilePath, Bool),
+    latestTypecheckedFile :: Maybe (UF.TypecheckedUnisonFile Symbol Ann),
+    -- The previous user input. Used to request confirmation of
+    -- questionable user commands.
+    lastInput :: Maybe Input,
+    -- A 1-indexed list of strings that can be referenced by index at the
+    -- CLI prompt.  e.g. Given ["Foo.bat", "Foo.cat"],
+    -- `rename 2 Foo.foo` will rename `Foo.cat` to `Foo.foo`.
+    numberedArgs :: NumberedArgs
+  }
+  deriving stock (Generic)
+
+instance
+  {-# OVERLAPS #-}
+  Functor f =>
+  IsLabel "currentPath" ((Path.Absolute -> f Path.Absolute) -> (LoopState -> f LoopState))
+  where
+  fromLabel :: (Path.Absolute -> f Path.Absolute) -> (LoopState -> f LoopState)
+  fromLabel =
+    lens
+      (\LoopState {currentPathStack} -> List.NonEmpty.head currentPathStack)
+      ( \loopState@LoopState {currentPathStack = _ List.NonEmpty.:| paths} path ->
+          loopState {currentPathStack = path List.NonEmpty.:| paths}
+      )
+
+-- | Create an initial loop state given a root branch and the current path.
+loopState0 :: Branch IO -> Path.Absolute -> LoopState
+loopState0 b p =
+  LoopState
+    { root = b,
+      lastSavedRoot = b,
+      currentPathStack = pure p,
+      latestFile = Nothing,
+      latestTypecheckedFile = Nothing,
+      lastInput = Nothing,
+      numberedArgs = []
+    }
+
+-- | Run a @Cli@ action down to @IO@.
+runCli :: Env -> Cli a a -> IO (ReturnType a)
+runCli env (Cli action) =
+  action env (\x -> pure (Success x))
+
+-- | Get the loop state.
+getLoopState :: Cli r LoopState
+getLoopState = do
+  Env {loopStateRef} <- ask
+  liftIO (readIORef loopStateRef)
+
+-- | Put the loop state.
+putLoopState :: LoopState -> Cli r ()
+putLoopState newSt = do
+  Env {loopStateRef} <- ask
+  liftIO (writeIORef loopStateRef newSt)
+
+-- | Modify the loop state (not thread-safe).
```

Comment:
I'm curious, what's the benefit of using an IORef instead of State if we're not aiming to use multiple threads or have MonadUnliftIO? 

Seems like it could be a foot-gun in the future.

If it's due to weird interactions between ContT and State I agree those are a pain, but I think we could work through them 🤔